packages/cursorless-vscode-e2e/src/suite/performance.vscode.test.ts

```diff
@@ -0,0 +1,140 @@
+import { endToEndTestSetup } from "../endToEndTestSetup";
+
+const testData = generateTestData();
+const numLines = testData.split("\n").length;
+
+suite(`Performance: ${numLines} lines JSON`, async function () {
+  endToEndTestSetup(this);
+
+  let previousTitle = "";
+
+  this.beforeEach(function () {
+    const title = this.currentTest!.title;
+    if (title !== previousTitle) {
+      console.log(`    ${title}`);
+      previousTitle = title;
+    }
+  });
+
+  const textBasedThreshold = 100;
+  const parseTreeThreshold = 500;
+  const surroundingPairThreshold = 30000;
+
+  test(
+    "Remove token",
+    asyncSafety(() => removeToken(textBasedThreshold)),
+  );
+
+  const fixtures: [SimpleScopeTypeType | ScopeType, number][] = [
+    // Text based
+    ["character", textBasedThreshold],
+    ["word", textBasedThreshold],
+    ["token", textBasedThreshold],
+    ["identifier", textBasedThreshold],
+    ["line", textBasedThreshold],
+    ["sentence", textBasedThreshold],
+    ["paragraph", textBasedThreshold],
+    ["document", textBasedThreshold],
+    ["nonWhitespaceSequence", textBasedThreshold],
+    // Parse tree based
+    ["string", parseTreeThreshold],
+    ["map", parseTreeThreshold],
+    ["collectionKey", parseTreeThreshold],
+    ["value", parseTreeThreshold],
+    // Text based, but utilizes surrounding pair
+    ["boundedParagraph", surroundingPairThreshold],
+    ["boundedNonWhitespaceSequence", surroundingPairThreshold],
+    ["collectionItem", surroundingPairThreshold],
+    // Surrounding pair
+    [{ type: "surroundingPair", delimiter: "any" }, surroundingPairThreshold],
+    [
+      { type: "surroundingPair", delimiter: "curlyBrackets" },
+      surroundingPairThreshold,
+    ],
+  ];
+
+  for (const [scope, threshold] of fixtures) {
+    const [scopeType, title] = getScopeTypeAndTitle(scope);
+    test(
+      `Select ${title}`,
+      asyncSafety(() => selectScopeType(scopeType, threshold)),
+    );
+  }
+});
+
+async function removeToken(threshold: number) {
+  await testPerformance(threshold, {
+    name: "remove",
+    target: {
+      type: "primitive",
+      modifiers: [{ type: "containingScope", scopeType: { type: "token" } }],
+    },
+  });
+}
+
+async function selectScopeType(scopeType: ScopeType, threshold: number) {
+  await testPerformance(threshold, {
+    name: "setSelection",
+    target: {
+      type: "primitive",
+      modifiers: [{ type: "containingScope", scopeType }],
+    },
+  });
+}
+
+async function testPerformance(threshold: number, action: ActionDescriptor) {
+  const editor = await openNewEditor(testData, { languageId: "json" });
+  const position = new vscode.Position(editor.document.lineCount - 3, 5);
+  const selection = new vscode.Selection(position, position);
+  editor.selections = [selection];
+  editor.revealRange(selection);
+
+  const start = performance.now();
+
+  await runCursorlessCommand({
+    version: 7,
+    usePrePhraseSnapshot: false,
+    action,
+  });
```

Comment:
I would be tempted to do maybe 3 runs and average? This could fail for example due to contention on the github actions CI server or something.

Reply:
I could do that. I kinda figured that since we will rerun a failing test it only takes one that's fast enough.

Reply:
Yeah that's fair. Let's leave it and if we see it start to cause CI failures we can revisit